numpy/_core/src/multiarray/item_selection.c

```diff
@@ -231,21 +231,32 @@ PyArray_TakeFrom(PyArrayObject *self0, PyObject *indices0, int axis,
                  PyArrayObject *out, NPY_CLIPMODE clipmode)
 {
     PyArray_Descr *dtype;
-    PyArrayObject *obj = NULL, *self, *indices;
+    PyArrayObject *obj = NULL, *self, *indices, *tmp;
     npy_intp nd, i, n, m, max_item, chunk, itemsize, nelem;
     npy_intp shape[NPY_MAXDIMS];
 
     npy_bool needs_refcounting;
 
-    indices = NULL;
+    indices = tmp = NULL;
     self = (PyArrayObject *)PyArray_CheckAxis(self0, &axis,
                                     NPY_ARRAY_CARRAY_RO);
     if (self == NULL) {
         return NULL;
     }
-    indices = (PyArrayObject *)PyArray_ContiguousFromAny(indices0,
-                                                         NPY_INTP,
-                                                         0, 0);
+    tmp = (PyArrayObject *)PyArray_ContiguousFromAny(indices0,
+                                                     NPY_INT64,
+                                                     0, 0);
```

Comment:
I wonder if it would be better to add a new array creation flag that tells `PyArray_FromArray` to use same-kind casting. Then this becomes a one-line change: you'd tell it to cast to INTP instead of INT64 here. 

There's `NPY_ARRAY_FORCECAST` to force it to use unsafe casting, but I don't think we want that.

There are internal-only array creation flags here, so you don't need to make this new flag public (although maybe it makes sense? Probably not worth the effort though):

https://github.com/numpy/numpy/blob/4ecffd70d6bf67f5e9533b258c0a17aa3c373e85/numpy/_core/src/multiarray/arrayobject.h#L36-L56

You'd add a new if statement here:

https://github.com/numpy/numpy/blob/4ecffd70d6bf67f5e9533b258c0a17aa3c373e85/numpy/_core/src/multiarray/ctors.c#L1911-L1914

I'd of course defer to @seberg on the best approach.

Reply:
I think in practice, a few places convert to array first (since if it isn't an array, we don't care about losing speed also) use force-cast and manually check the same-kind casting.
I don't mind adding a flag, but would have to think about it for the public API.